didcomm/core/utils.py

```diff
@@ -166,38 +169,136 @@ def _extract_key_from_secret(secret: Secret, align_kid) -> AsymmetricKey:
+        raw_x_value = raw_value[_CURVE25519_POINT_SIZE:]
+
+        base64url_d_value = urlsafe_b64encode(raw_d_value)
+        base64url_x_value = urlsafe_b64encode(raw_x_value)
+
+        jwk = {
+            "kty": "OKP",
+            "crv": "X25519"
+            if secret.type == VerificationMethodType.X25519_KEY_AGREEMENT_KEY_2019
+            else "Ed25519",
+            "x": to_unicode(base64url_x_value),
+            "d": to_unicode(base64url_d_value),
+        }
+
+        if align_kid:
+            jwk["kid"] = secret.kid
+
+        return OKPKey.import_key(jwk)
+
+    elif secret.type in [
+        VerificationMethodType.X25519_KEY_AGREEMENT_KEY_2020,
+        VerificationMethodType.ED25519_VERIFICATION_KEY_2020,
+    ]:
+        if secret.verification_material.format != VerificationMaterialFormat.MULTIBASE:
+            raise DIDCommValueError()
+
+        # Currently only base58btc encoding is supported in scope of multibase support
+        if secret.verification_material.value.startswith("z"):
+            prefixed_raw_value = base58.b58decode(
+                secret.verification_material.value[1:]
+            )
+        else:
+            raise DIDCommValueError()
+
+        codec, raw_value = _from_multicodec(prefixed_raw_value)
+
+        expected_codec = (
+            _Codec.X25519_PRIV
+            if secret.type == VerificationMethodType.X25519_KEY_AGREEMENT_KEY_2020
+            else _Codec.ED25519_PRIV
+        )
+
+        if codec != expected_codec:
+            raise DIDCommValueError()
+
+        raw_d_value = raw_value[:_CURVE25519_POINT_SIZE]
+        raw_x_value = raw_value[_CURVE25519_POINT_SIZE:]
+
+        base64url_d_value = urlsafe_b64encode(raw_d_value)
+        base64url_x_value = urlsafe_b64encode(raw_x_value)
+
+        jwk = {
+            "kty": "OKP",
+            "crv": "X25519"
+            if secret.type == VerificationMethodType.X25519_KEY_AGREEMENT_KEY_2020
+            else "Ed25519",
+            "x": to_unicode(base64url_x_value),
+            "d": to_unicode(base64url_d_value),
+        }
+
+        if align_kid:
+            jwk["kid"] = secret.kid
+
+        return OKPKey.import_key(jwk)
+
     else:
         raise DIDCommValueError()
 
 
-def extract_sign_alg(verification_method: Union[VerificationMethod, Secret]) -> SignAlg:
-    if (
-        verification_method.type == VerificationMethodType.JSON_WEB_KEY_2020
-        and verification_method.verification_material.format
-        == VerificationMaterialFormat.JWK
-    ):
-        jwk = json_str_to_dict(verification_method.verification_material.value)
+class _Codec(Enum):
+    X25519_PUB = 0xEC
+    ED25519_PUB = 0xED
+    ED25519_PRIV = 0x1300
+    X25519_PRIV = 0x1302
+
+
+def _from_multicodec(value: bytes) -> (_Codec, bytes):
+    try:
+        prefix_int = varint.decode_bytes(value)
+    except Exception:
+        raise ValueError("Invalid multicodec prefix in {}".format(str(value)))
```

Comment:
I think it should be `DIDCommValueError`.

Reply:
Done.